# 14. Início da Modelagem Preditiva

Nesta etapa iniciamos a preparação dos dados para Machine Learning.  
Carregamos novamente as bases (2020 e 2021), criamos cópias para preservar os dados brutos  
e verificamos a estrutura das tabelas para identificar tipos de variáveis e possíveis ajustes.


In [91]:
# Importação de bibliotecas principais para ML
import pandas as pd
import numpy as np

# Funções úteis do scikit-learn para treino e avaliação
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder 
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Modelos de regressão
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.ensemble import ExtraTreesRegressor, GradientBoostingRegressor


In [92]:
# Leitura das bases CSV
raw_2021 = pd.read_csv('data/ifood-restaurants-february-2021.csv')
raw_2020 = pd.read_csv('data/ifood-restaurants-november-2020.csv')

In [93]:
# Criação de cópias para não alterar os dados originais
df_2021 = raw_2021.copy()
df_2020 = raw_2020.copy()

In [94]:
# Estrutura da base de 2020
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 361447 entries, 0 to 361446
Data columns (total 11 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   avatar         361149 non-null  object 
 1   category       361447 non-null  object 
 2   delivery_fee   361447 non-null  float64
 3   delivery_time  361447 non-null  int64  
 4   distance       361447 non-null  float64
 5   name           361447 non-null  object 
 6   price_range    361447 non-null  object 
 7   rating         361447 non-null  float64
 8   url            361447 non-null  object 
 9   lat            361447 non-null  float64
 10  long           361447 non-null  float64
dtypes: float64(5), int64(1), object(5)
memory usage: 30.3+ MB


In [95]:
# Estrutura da base de 2020
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 361447 entries, 0 to 361446
Data columns (total 11 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   avatar         361149 non-null  object 
 1   category       361447 non-null  object 
 2   delivery_fee   361447 non-null  float64
 3   delivery_time  361447 non-null  int64  
 4   distance       361447 non-null  float64
 5   name           361447 non-null  object 
 6   price_range    361447 non-null  object 
 7   rating         361447 non-null  float64
 8   url            361447 non-null  object 
 9   lat            361447 non-null  float64
 10  long           361447 non-null  float64
dtypes: float64(5), int64(1), object(5)
memory usage: 30.3+ MB


In [96]:
# Estrutura da base de 2021
df_2021.head(4)

availableForScheduling                                             avatar  \
0                   False  https://static-images.ifood.com.br/image/uploa...   
1                   False  https://static-images.ifood.com.br/image/uploa...   
2                   False  https://static-images.ifood.com.br/image/uploa...   
3                   False  https://static-images.ifood.com.br/image/uploa...   

  category  delivery_fee  delivery_time  distance     ibge  minimumOrderValue  \
0  Marmita          3.99             27      1.22  5300108               10.0   
1     Açaí          7.99             61      4.96  5300108               10.0   
2  Bebidas         11.99             70      8.35  5300108                5.0   
3   Carnes         16.49             63      6.35  5300108               20.0   

                         name  \
0        Cantina Arte & Sabor   
1            Raruty Açaí Raiz   
2               Toma na Kombi   
3  Churrasquinho do Barriga´s   

                                        paymentCodes price_range  rating  \
0  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   
1  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   
2  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ R...    MODERATE     0.0   
3  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   

                                                tags  \
0  ADDRESS_PREFORM_TYPE $$ CART::MCHT::100_DELIVE...   
1  ADDRESS_PREFORM_TYPE $$ GUIDED_HELP_TYPE $$ ME...   
2  ADDRESS_PREFORM_TYPE $$ CPGN_USER_DISCOUNT_6_L...   
3  ADDRESS_PREFORM_TYPE $$ GUIDED_HELP_TYPE $$ NO...   

                                                 url  
0  https://www.ifood.com.br/delivery/brasilia-df/...  
1  https://www.ifood.com.br/delivery/brasilia-df/...  
2  https://www.ifood.com.br/delivery/brasilia-df/...  
3  https://www.ifood.com.br/delivery/brasilia-df/...

## 15. Estruturação e Preparação dos Dados

- Padronização das colunas de 2020 e 2021.  
- Criação da coluna `state` a partir da URL.  
- Remoção de variáveis irrelevantes (ex.: `avatar`, `url`, `tags`).  
- Inclusão de coluna `ano` para diferenciar os datasets.  
- União das bases em um único dataframe (`df_total`) para análise preditiva.  


In [97]:
# Extrai o 'state' da URL em 2020
df_2020['state'] = df_2020['url'].str.split("/").str[4].str.split("-").str[-1]

In [98]:
# Remove colunas irrelevantes da base de 2020
df_2020 = df_2020.drop(columns=['avatar', 'url', 'lat', 'long'])

In [99]:
# Visualiza primeiras linhas de 2021
df_2021.head(4)

availableForScheduling                                             avatar  \
0                   False  https://static-images.ifood.com.br/image/uploa...   
1                   False  https://static-images.ifood.com.br/image/uploa...   
2                   False  https://static-images.ifood.com.br/image/uploa...   
3                   False  https://static-images.ifood.com.br/image/uploa...   

  category  delivery_fee  delivery_time  distance     ibge  minimumOrderValue  \
0  Marmita          3.99             27      1.22  5300108               10.0   
1     Açaí          7.99             61      4.96  5300108               10.0   
2  Bebidas         11.99             70      8.35  5300108                5.0   
3   Carnes         16.49             63      6.35  5300108               20.0   

                         name  \
0        Cantina Arte & Sabor   
1            Raruty Açaí Raiz   
2               Toma na Kombi   
3  Churrasquinho do Barriga´s   

                                        paymentCodes price_range  rating  \
0  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   
1  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   
2  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ R...    MODERATE     0.0   
3  DNR $$ MPAY $$ MOVPAY_MC $$ MC $$ GPY_ELO $$ E...    CHEAPEST     0.0   

                                                tags  \
0  ADDRESS_PREFORM_TYPE $$ CART::MCHT::100_DELIVE...   
1  ADDRESS_PREFORM_TYPE $$ GUIDED_HELP_TYPE $$ ME...   
2  ADDRESS_PREFORM_TYPE $$ CPGN_USER_DISCOUNT_6_L...   
3  ADDRESS_PREFORM_TYPE $$ GUIDED_HELP_TYPE $$ NO...   

                                                 url  
0  https://www.ifood.com.br/delivery/brasilia-df/...  
1  https://www.ifood.com.br/delivery/brasilia-df/...  
2  https://www.ifood.com.br/delivery/brasilia-df/...  
3  https://www.ifood.com.br/delivery/brasilia-df/...

In [100]:
# Extrai o 'state' da URL em 2021
df_2021['state'] = df_2021['url'].str.split("/").str[4].str.split("-").str[-1]

In [101]:
# Remove colunas irrelevantes da base de 2021
df_2021 = df_2021.drop(columns=['availableForScheduling', 'avatar', 'ibge', 'paymentCodes', 'tags', 'url'])

In [102]:
# Cria coluna 'ano' para diferenciar os datasets
df_2020['ano'] = 2020
df_2021['ano'] = 2021

In [103]:
df_2021.tail(8)

category  delivery_fee  delivery_time  distance  \
406391        Lanches           3.0             50      1.77   
406392        Lanches           3.5             60      2.74   
406393        Lanches           0.0             60      2.58   
406394           Açaí           9.0             60      3.53   
406395           Açaí           6.0             50      2.60   
406396           Açaí           0.0             40      3.61   
406397        Lanches           8.0             60      3.54   
406398  Doces & Bolos           9.0             20      0.95   

        minimumOrderValue                               name price_range  \
406391               20.0                Hamburgueria Duarte    MODERATE   
406392               15.0                           ki-sabor       CHEAP   
406393               15.0                Pizza Cone del Miko    CHEAPEST   
406394               30.0                       Açaí da Duda    CHEAPEST   
406395               10.0                      Pede aí açaí     CHEAPEST   
406396                0.0        Açaí do Jeitinho Brasileiro    CHEAPEST   
406397               20.0                     Classic Burger    CHEAPEST   
406398                0.0  Cacau Show - Cachoerinha Shopping    MODERATE   

         rating state   ano  
406391  5.00000    rs  2021  
406392  5.00000    rs  2021  
406393  5.00000    rs  2021  
406394  4.95000    rs  2021  
406395  0.00000    rs  2021  
406396  4.46602    rs  2021  
406397  5.00000    rs  2021  
406398  4.84000    rs  2021

## 14. Ajuste da coluna `minimumOrderValue`

- Em 2021 já existe a coluna `minimumOrderValue`.  
- Para 2020, os valores foram atribuídos a partir da correspondência com `delivery_fee` (mapeando pela mediana).  
- Valores que não tinham correspondência foram preenchidos com a mediana geral de 2021.  
- Ao final, as bases de 2020 e 2021 foram consolidadas em um único dataframe (`df_total`).


In [104]:
# Criar tabela de correspondência a partir de 2021
# (se tiver mais de um minimumOrderValue para a mesma taxa, usamos a média)
map_min_order = (
    df_2021.groupby('delivery_fee')['minimumOrderValue']
    .mean()
    .to_dict())

# Aplicar regra no 2020
df_2020['minimumOrderValue'] = df_2020['delivery_fee'].map(map_min_order)

# Preenche valores ausentes de 2020 (sem correspondência) com a média geral de 2021
df_2020['minimumOrderValue'] = df_2020['minimumOrderValue'].fillna(df_2021['minimumOrderValue'].mean())


In [105]:
# Visualiza últimas linhas do dataset 2020 já tratado
df_2020.tail(8)

category  delivery_fee  delivery_time  distance  \
361439     Brasileira          3.99             38      1.82   
361440        Bebidas          3.99             42      1.74   
361441      Argentina          3.99             38      1.78   
361442     Brasileira          3.99             18      1.66   
361443        Lanches          3.99             18      1.70   
361444          Pizza          8.00             45      1.34   
361445      Cafeteria          1.00             35      0.20   
361446  Doces & Bolos          0.00             60      0.96   

                                           name price_range  rating state  \
361439            Restaurante Cantinho da Serra    MODERATE    0.00    sp   
361440                           Boteco Itapeva    MODERATE    4.75    sp   
361441                  Cantinho da Serra Grill    MODERATE    0.00    sp   
361442                Restaurante Nosso Tempero    CHEAPEST    0.00    sp   
361443                          Belgian Waffles    CHEAPEST    0.00    sp   
361444   Sabor de Campos Pizzaria e Restaurante    CHEAPEST    0.00    sp   
361445                             Maria Bonita    CHEAPEST    0.00    sc   
361446  Cacau Show - Supermercado Campos Salles    CHEAPEST    5.00    ce   

         ano  minimumOrderValue  
361439  2020          15.429268  
361440  2020          15.429268  
361441  2020          15.429268  
361442  2020          15.429268  
361443  2020          15.429268  
361444  2020          14.257010  
361445  2020          11.510230  
361446  2020          21.016338

In [106]:
# Junta bases 2020 e 2021
df_total = pd.concat([df_2020, df_2021], ignore_index=True)

In [107]:
# Mostra dimensão da base consolidada
print(df_total.shape)


(767846, 10)


In [108]:
# Visualiza primeiras linhas do dataset consolidado
df_total.head()

category  delivery_fee  delivery_time  distance  \
0        Lanches          9.00             80      6.20   
1  Doces & Bolos          6.00             35      3.03   
2     Brasileira          4.00             40      1.51   
3        Lanches          0.00             50      0.79   
4       Salgados         12.99             36      5.12   

                                    name price_range   rating state   ano  \
0                              El'moedor       CHEAP  4.30303    pe  2020   
1                  Delicia de Brigadeiro    CHEAPEST  0.00000    pe  2020   
2              Pizzaria Rappi10 - Moreno    CHEAPEST  0.00000    pe  2020   
3                       Tapioca Arretada    CHEAPEST  5.00000    pe  2020   
4  Minuto Kit Festa ( Salgados e Doces )    CHEAPEST  5.00000    pe  2020   

   minimumOrderValue  
0          13.483475  
1          13.599419  
2         366.540819  
3          21.016338  
4          15.018331

## 15. Tratamento da variável alvo (`rating`)

- Conversão da coluna `rating` para tipo numérico.  
- Remoção de valores inválidos (fora do intervalo 1–5).  
- Exclusão de linhas sem avaliação (`NaN`).  
- Análise do impacto da limpeza na quantidade de registros.  


In [109]:
# Converte 'rating' para numérico (se houver erro, vira NaN)
df_total['rating'] = pd.to_numeric(df_total['rating'], errors='coerce')

# Substitui avaliações iguais a 0 por NaN (sem avaliação)
df_total.loc[df_total['rating'] == 0, 'rating'] = pd.NA

# manter apenas valores entre 1 e 5
df_total = df_total[df_total['rating'].between(1, 5, inclusive='both')]


In [110]:
# Estatísticas descritivas do rating após limpeza
df_total['rating'].describe()

count    448237.000000
mean          4.518793
std           0.604238
min           1.000000
25%           4.363640
50%           4.667880
75%           4.916670
max           5.000000
Name: rating, dtype: float64

In [111]:
# Conta valores ausentes em 'rating'
print("Qtd de NaN em rating:", df_total['rating'].isna().sum())


Qtd de NaN em rating: 0


In [112]:
# Avalia impacto da limpeza no tamanho da base
total_original = 767846
total_final = df_total.shape[0]

removidos = total_original - total_final
perc = removidos / total_original * 100

print(f"Total original: {total_original:,}")
print(f"Após limpeza: {total_final:,}")
print(f"Linhas removidas (rating = 0/NaN): {removidos:,}. ({perc:.1f}%).")


Total original: 767,846
Após limpeza: 448,237
Linhas removidas (rating = 0/NaN): 319,609. (41.6%).


## 16. Tratamento de outliers e preparação das variáveis

- Aplicação de **regras de corte** para limitar valores extremos em `delivery_time`, `distance` e `minimumOrderValue`.  
- Definição das variáveis independentes (X) e da variável alvo (`rating`).  
- Separação de colunas em **categóricas** e **numéricas**.  
- Aplicação de **One-Hot Encoding** para transformar variáveis categóricas em formato numérico.  
- Consolidação do dataset final (`X_final`) pronto para treino e teste dos modelos.  


In [113]:
# Estatísticas das variáveis numéricas relevantes
num_cols = ['delivery_fee','delivery_time','distance','minimumOrderValue','ano']
print(df_total[num_cols].describe().T)

                      count         mean            std      min      25%  \
delivery_fee       448237.0     6.682139       4.357669     0.00     4.00   
delivery_time      448237.0    46.215730      20.929946    -5.00    34.00   
distance           448237.0     3.529183      33.750311     0.01     1.71   
minimumOrderValue  448237.0   374.414960  149366.793103     0.00    13.00   
ano                448237.0  2020.506957       0.499952  2020.00  2020.00   

                           50%      75%          max  
delivery_fee          6.000000     9.00        40.00  
delivery_time        45.000000    60.00      5060.00  
distance              3.050000     4.77     11170.24  
minimumOrderValue    15.429268    20.00  99999999.99  
ano                2021.000000  2021.00      2021.00  


In [114]:
# olhar valores máximos por coluna
for c in num_cols:
    print(f"\nColuna: {c}")
    print("Top maiores valores:")
    print(df_total[c].sort_values(ascending=False).head(5).to_list())


Coluna: delivery_fee
Top maiores valores:
[40.0, 40.0, 36.9, 35.0, 35.0]

Coluna: delivery_time
Top maiores valores:
[5060, 5050, 1335, 480, 350]

Coluna: distance
Top maiores valores:
[11170.24, 11170.24, 9843.84, 9274.69, 5196.45]

Coluna: minimumOrderValue
Top maiores valores:
[99999999.99, 150000.0, 100000.0, 100000.0, 30000.0]

Coluna: ano
Top maiores valores:
[2021, 2021, 2021, 2021, 2021]


In [115]:
# regras de corte
df_total = df_total[df_total['delivery_time'] <= 180]
df_total = df_total[df_total['distance'] <= 50]
df_total = df_total[df_total['minimumOrderValue'] <= 50]

# Verifica tamanho da base após remoção dos outliers
print("Novo shape:", df_total.shape)
print(df_total[['delivery_fee','delivery_time','distance','minimumOrderValue']].describe())


Novo shape: (421420, 10)
        delivery_fee  delivery_time       distance  minimumOrderValue
count  421420.000000  421420.000000  421420.000000      421420.000000
mean        6.775921      46.026731       3.404854          15.789975
std         4.420967      17.874158       2.125572           6.686551
min         0.000000      -5.000000       0.010000           0.000000
25%         4.000000      33.000000       1.720000          11.764267
50%         6.000000      45.000000       3.070000          15.286707
75%         9.490000      60.000000       4.860000          20.000000
max        35.000000     180.000000      43.570000          50.000000


In [116]:
# Definição das variáveis
cat_cols = ['category', 'price_range', 'state']
num_cols = ['delivery_fee', 'delivery_time', 'distance', 'minimumOrderValue', 'ano']

X = df_total[cat_cols + num_cols].copy()
y = df_total['rating'].astype(float)

# One-Hot Encoding nas variáveis categóricas 
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
X_encoded = encoder.fit_transform(X[cat_cols])

# Cria DataFrame com os nomes das colunas codificadas
encoded_cols = encoder.get_feature_names_out(cat_cols)
X_encoded = pd.DataFrame(X_encoded, columns=encoded_cols, index=X.index)

# Junta numéricas + categóricas codificadas
X_final = pd.concat([X[num_cols], X_encoded], axis=1)

# Visualiza primeiras linhas da base final para modelagem
print(X_final.shape)
X_final.head(3)

(421420, 97)


delivery_fee  delivery_time  distance  minimumOrderValue   ano  \
0          9.00             80      6.20          13.483475  2020   
3          0.00             50      0.79          21.016338  2020   
4         12.99             36      5.12          15.018331  2020   

   category_Africana  category_Alemã  category_Argentina  category_Asiática  \
0                0.0             0.0                 0.0                0.0   
3                0.0             0.0                 0.0                0.0   
4                0.0             0.0                 0.0                0.0   

   category_Açaí  ...  state_pr  state_rj  state_rn  state_ro  state_rr  \
0            0.0  ...       0.0       0.0       0.0       0.0       0.0   
3            0.0  ...       0.0       0.0       0.0       0.0       0.0   
4            0.0  ...       0.0       0.0       0.0       0.0       0.0   

   state_rs  state_sc  state_se  state_sp  state_to  
0       0.0       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0       0.0  

[3 rows x 97 columns]

## 17. Treinamento e Avaliação de Modelos

- Divisão da base em treino (80%) e teste (20%).  
- Execução de diferentes algoritmos:  
  - **Linear Regression** (baseline).  
  - **Random Forest** (árvores em conjunto).  
  - **Extra Trees**, **Gradient Boosting** e **XGBoost** (modelos de ensemble).  
- Avaliação com métricas: **MAE (erro absoluto médio)**, **RMSE (raiz do erro quadrático médio)** e **R² (coeficiente de determinação)**.  
- Comparação de desempenho entre os modelos.  


In [117]:
# Split dos dados em treino e teste 
X_train, X_test, y_train, y_test = train_test_split(X_final, y, test_size=0.2, random_state=42)

# 1) Linear Regression
lin = LinearRegression().fit(X_train, y_train)
p_lin = lin.predict(X_test)

mae_lin  = mean_absolute_error(y_test, p_lin)
rmse_lin = np.sqrt(mean_squared_error(y_test, p_lin))
r2_lin   = r2_score(y_test, p_lin)

# 2) Random Forest (modelo não linear baseado em árvores)
rf = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1).fit(X_train, y_train)
p_rf = rf.predict(X_test)

# Comparação inicial Linear x Random Forest
mae_rf  = mean_absolute_error(y_test, p_rf)
rmse_rf = np.sqrt(mean_squared_error(y_test, p_rf))
r2_rf   = r2_score(y_test, p_rf)

# Comparação inicial Linear x Random Forest
results = pd.DataFrame(
    [["Linear Regression", mae_lin, rmse_lin, r2_lin],
    ["Random Forest",    mae_rf,  rmse_rf,  r2_rf ],],
    columns=["Modelo", "MAE", "RMSE", "R²"]).round({"MAE":3, "RMSE":3, "R²":3})

print(results)

              Modelo    MAE   RMSE     R²
0  Linear Regression  0.379  0.592  0.039
1      Random Forest  0.399  0.619 -0.050


In [118]:
# 3) Modelos adicionais (ExtraTrees, GradientBoosting, XGBoost)
models = {
    "ExtraTrees": ExtraTreesRegressor(n_estimators=200, random_state=42, n_jobs=-1),
    "GradientBoosting": GradientBoostingRegressor(n_estimators=200, random_state=42),
    "XGBoost": XGBRegressor(n_estimators=200, learning_rate=0.1, max_depth=8, random_state=42, n_jobs=-1, tree_method="hist"),
}

results = []

for name, model in models.items():
    model.fit(X_train, y_train)
    pred = model.predict(X_test)

    mae = mean_absolute_error(y_test, pred)
    rmse = np.sqrt(mean_squared_error(y_test, pred))
    r2 = r2_score(y_test, pred)

    results.append([name, mae, rmse, r2])
    
# Comparação final dos modelos ensemble
df_results = pd.DataFrame(results, columns=["Modelo","MAE","RMSE","R²"]).round(4)
print(df_results)


             Modelo     MAE    RMSE      R²
0        ExtraTrees  0.4265  0.6749 -0.2494
1  GradientBoosting  0.3770  0.5898  0.0458
2           XGBoost  0.3743  0.5875  0.0534


## 19. Transformação do `rating` em classes

- Criação de grupos de avaliação a partir do `rating`:  
  - **Ruim (1–2)**  
  - **Neutro (3)**  
  - **Bom (4–5)**  
- Nova coluna `rating_group` adicionada ao dataset.  
- Cálculo da distribuição percentual de cada classe para entender o balanceamento.  

In [119]:
# Função para classificar notas de rating em grupos
def rating_group(r):
    if r <= 2:
        return "Ruim (1-2)"
    elif r == 3:
        return "Neutro (3)"
    else:
        return "Bom (4-5)"
    
# Cria nova coluna com as classes de avaliação
df_total["rating_group"] = df_total["rating"].apply(rating_group)

# Contagem e proporção
dist = df_total["rating_group"].value_counts(normalize=True).mul(100).round(2)
print(dist)


rating_group
Bom (4-5)     97.16
Ruim (1-2)     1.46
Neutro (3)     1.38
Name: proportion, dtype: float64


---
## 📌 Conclusão do Projeto de Machine Learning (iFood)

### 1. O que foi feito. 
- Realizamos a **preparação das bases de 2020 e 2021**, padronizando colunas, criando identificadores (`id`, `state`, `city`, `ano`) e removendo atributos irrelevantes.  
- Fizemos a **limpeza do target `rating`**, transformando notas `0` em `NaN` e mantendo apenas valores entre **1 e 5**.  
- Tratamos **outliers** em `delivery_time`, `distance` e `minimumOrderValue`.  
- Criamos a matriz final com **variáveis numéricas e categóricas (via OneHotEncoder)**.  
- Testamos diferentes modelos de regressão (**Linear, Random Forest, ExtraTrees, GradientBoosting, XGBoost**) avaliando com **MAE, RMSE e R²**.  

### 2. Por que usamos `rating` como alvo.
- A princípio, a ideia era prever a **avaliação média dos restaurantes** a partir de características operacionais (frete, tempo de entrega, distância, preço etc.).  
- Isso faria sentido como métrica de **qualidade percebida pelo cliente**.  

### 3. O que encontramos.
- Apesar de um erro médio relativamente baixo (~0.38 estrelas), o **R² ficou muito baixo (≈5–7%)** → ou seja, os modelos **não conseguem explicar a variabilidade** das notas.  
- Isso acontece porque o `rating` está **extremamente desbalanceado**:  
- ~97% dos restaurantes têm nota **4 ou 5**  
- Notas baixas (1–2) são quase inexistentes.  
- Além disso, o `rating` depende de fatores **fora da base** (qualidade da comida, atendimento, promoções, marketing), que não conseguimos capturar com as features disponíveis.  

### 4. Conclusão.
- O problema **não é o modelo**, mas o **alvo escolhido**.  
- Com os dados disponíveis, **prever `rating` não é viável**, pois falta variabilidade e explicação estatística.  
- O pipeline continua válido: demonstramos todo o processo de **EDA, feature engineering, preparação, modelagem e avaliação**, mas justificamos o **encerramento da modelagem preditiva com `rating`**.  

### 5. Próximos passos recomendados.  
- Substituir o alvo por variáveis mais explicáveis, como:  
- `delivery_fee` (predição de custo de entrega)  
- `delivery_time` (predição de tempo estimado)  
- Popularidade (`número de avaliações`).  
- Insistir na predição do `rating` com os dados atuais resultaria em **modelos pouco explicativos**, que no máximo fariam um "chute sofisticado" sempre prevendo próximo de 4 ou 5.  
- Isso geraria **falsos insights** para o negócio, já que o modelo não estaria de fato aprendendo padrões relevantes.  
- Continuar nesse caminho traria risco de apresentar resultados **enganosos** para stakeholders, sem ganho prático.  

---
**Resumo final.**  
- O projeto demonstrou todo o pipeline de ML (preparo, tratamento, modelagem, avaliação).  
- Porém, ficou claro que **o `rating` não é adequado como variável alvo** com as features disponíveis.  
- Assim, o trabalho é **encerrado aqui de forma consciente**, justificando que **um bom modelo exige um alvo de qualidade**.  
- Caso novos dados sejam incorporados ou o problema seja redefinido, o pipeline já está pronto para ser reaproveitado.


---